### PR TITLE
Refactor storage operations into separate Backend classes

### PR DESCRIPTION
**Pull request by elfjes**
Following the discussion in #253 and #325 I've created a first iteration on what a `Backend` interface could look like and how the current file storage operations may be refactored into this interface. It goes from the following principles

* `app.py` talks only to `core.py` with regards to package operations
* at configuration time, a `Backend` implementation is chosen and created for the lifetime of the configured app
* `core.py` proxies requests for packages to this `Backend()`
* The `Backend` interface/api is defined through three things
  * methods that an implementation must implement
  * methods that an implementation may override if it knows better than the defaults
  * the `PkgFIle` class that is (should be) the main carrier of data
* where possible, implementation details must be hidden from concrete `Backend`s to promote extensibility

Other things I've done in this PR:
* I've tried to talk about packages and projects, rather than files and prefixes, since these are the domain terms PEP503 uses, and imho it's also more clear what it means
* Better testability of the `CacheManager` (no more race conditions when `watchdog` is installed during testing)
* Cleanup some more Python 2 code
* Started moving away from  `os.path` and `py.path` in favour of `pathlib`

Furthermore I've created a `plugin.py` with a sample of how I think plugin system could look like. This sampIe assumes we use `argparse`  and allows for the extension of cli arguments that a plugin may need. I think the actual implementation of such a plugin system is beyond the scope of this PR, but I've used it as a target for the Backend refactoring. If requested, I'll remove it from this PR.

The following things still need to be done / discussed. These can be part of this PR or moved into their own, separate PRs
- [ ] Simplify the `PgkFile` class. It currently consists of a number of attributes that don't necessarily belong with it, and not all attributes are aptly named (imho). I would like to minimalize the scope of `PkgFile` so that its only concern is being a data carrier between the app and the backends, and make its use more clear.
- [ ] Add a `PkgFile.metadata` that backend implementations may use to store custom data for packages. For example the current `PkgFile.root` attribute is an implementation detail of the filestorage backends, and other Backend implementations should not be bothered by it.
- [ ] Use `pathlib` wherever possible. This may also result in less attributes for `PkgFile`, since some things may be just contained in a single `Path` object, instead of multtiple strings.
- [ ] Improve testing of the `CacheManager`.

### CODE REVIEW

**Comment by mplanchard** [comment]
Awesome! Just FYI I have what will probably be a long day of work ahead and
some plans for this evening, so I probably won’t get to review this until
Friday evening, but I’m looking forward to it!

**Comment by elfjes** [comment]
Sure no worries. Looking forward to your feedback :smile:

**Comment by mplanchard** [comment]
Got a chance to give this a quick overview this evening, and I really like the direction! I have an in-progress branch to replace all usage of the old config with the stuff from #339 with all tests passing. You can see that [here](https://github.com/pypiserver/pypiserver/compare/config-replace). It seems like that could mesh really well with the approach taken here. I think we would update the `RunConfig` to expose a `backend` attribute, rather than having it expose an `iter_packages` method.

I've got no major concerns with what I've seen so far, but I'll be sure to make time to give this a more thorough review over the course of this week.

In regards to the things you mentioned as potentially needing more work:

> Simplify the PgkFile class. It currently consists of a number of attributes that don't necessarily belong with it, and not all attributes are aptly named (imho). I would like to minimalize the scope of PkgFile so that its only concern is being a data carrier between the app and the backends, and make its use more clear.

I agree with this goal! I don't feel strongly either way on whether it should be a part of this PR, but I definitely think it's worth exploring

> Add a PkgFile.metadata that backend implementations may use to store custom data for packages. For example the current PkgFile.root attribute is an implementation detail of the filestorage backends, and other Backend implementations should not be bothered by it.

Yeah, and having some access to metadata will also give us a route towards e.g. fully supporting PEP 503 and the `data_requires` attribute (a la #320).

> Use pathlib wherever possible. This may also result in less attributes for PkgFile, since some things may be just contained in a single Path object, instead of multtiple strings.

I've been working in this direction in my branch as well, so definitely no issues with seeing more of it.

> Improve testing of the CacheManager.

Yeah the existing watchdog cache had really minimal tests, so anything here is an improvement

---

Regarding getting both of our streams of work going harmoniously, we can either get this merged, and then I can rebase and update on top of this, or vice-versa. I'll probably go ahead and open a PR with that branch tonight or tomorrow, just so that we can better look at how they might work together, but there's no rush to get that one in, and I'm happy to do the work of rebasing if we get this one in first.

File: pypiserver/core.py
Comment (mplanchard) [inline pypiserver/core.py:129]: to keep the decorator from overriding the underlying type, you can use a generic like:

```py
DigesterFunc = t.TypeVar("DigesterFunc", bound=t.Callable[..., t.Iterable[PkgFile]])

def with_digester(func: DigesterFunc) -> DigesterFunc:
  ...
```

you may need a `# type: ignore` where you're returning your wrapper function inside the decorator, but by telling mypy that you're returning a type of the same type that was put into the function, you allow it to infer that the resulting decorated function takes the same arguments as the original

File: pypiserver/manage.py
Comment (mplanchard) [inline pypiserver/manage.py:172]: I know you didn't write the original, but since we're touching it, we can keep this (potentially) lazy by updating like:

```suggestion
    all_packages = itertools.chain.from_iterable(listdir(r) for r in roots)
```

File: tests/test_app.py
Comment (mplanchard) [inline tests/test_app.py:116]: dead code?

File: pypiserver/_app.py
Comment (mplanchard) [inline pypiserver/_app.py:94]: It would be nice if we added a "package_count" or similar method to the backend so that, for example, we wouldn't actually have to parse all the files into `PkgFile` instances in order to get their count, and could instead just nab all files whose names match a regex or something. It would also be useful with imagined remote backends, where pulling all the files might be a problem

File: pypiserver/_app.py
Comment (mplanchard) [inline pypiserver/_app.py:333]: Could we try replacing this list comprehension (`[]`) with a generator comprehension (`()`)? We definitely need to allocate a list for sorting, but here I think we may be able to be lazy and just pass a generator into the template call, which should save some RAM on systems handling a lot of packages

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:71]: Could we explicitly say that this returns `t.Iterator[PkgFile]` instead of `t.Iterable`? I would like to guide implementations into being lazy with returning all the packages wherever possible

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:77]: bikeshed0: I think `stream` or `bytestream` may be a more generic name for this parameter and will make more sense for implementations not using the FS

we may also explicitly want to note the return type being `None` here (and for other methods with no return value)
```suggestion
    def add_package(self, filename: str, fh: t.BinaryIO) -> None:
```

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:90]:
```suggestion
    def exists(self, filename: str) -> bool:
```

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:99]: same note here re: iterator
```suggestion
    def get_projects(self) -> t.Iterator[str]:
```

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:108]: Is this check for the `pkgname` necessary? It seems like, since it' a required positional argument to the `PkgFile`, we should be able to rely on it (and the normalized name) always being present, in which case we could do something like

```py
yield from set(pkg.pkgname_norm for pkg in self.get_all_packages())
```

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:111]:
```suggestion
    def find_project_packages(self, project: str) -> t.Iterator[PkgFile]:
```

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:123]:
```suggestion
    def find_version(self, name, version) -> t.Iterator[PkgFile]:
```

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:67]: It would probably be helpful to implementors to include an example or two of what we expect each of these methods' inputs to be in their docstrings

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:173]: bikeshed: `as_file()` or `to_file()` feels to me like the function will be returning something file-like. What about `write_file()`?

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:150]: I think currently we're recursing into root directories, so we may want to do so here too

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:156]: We may want to define either a base class or a [protocol](https://docs.python.org/3/library/typing.html#typing.Protocol) for the cache_manager

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:135]: I think mypy will probably want us to define types for the subclasses. Since this is a new file, we can add it to the typechecks in `tox.ini` to make sure it catches anything we miss

File: pypiserver/backend.py
Comment (mplanchard) [inline pypiserver/backend.py:220]: let's pull in my update from #349, which essentially just replace this line with
```suggestion
    digester = hashlib.new(hash_algo)
```
to support a wider variety of hashing algorithms

**Review by mplanchard — COMMENTED** [review]
Found some time this evening for a first pass!

File: pypiserver/core.py
Comment (elfjes) [reply pypiserver/core.py:129]: good idea! :)

File: pypiserver/manage.py
Comment (elfjes) [reply pypiserver/manage.py:172]: Sure, I wanted to touch manage.py as little as possible, but this totally makes sense

File: tests/test_app.py
Comment (elfjes) [reply tests/test_app.py:116]: good catch, will remove

File: pypiserver/_app.py
Comment (elfjes) [reply pypiserver/_app.py:94]: a `package_count` method sounds good, but I'm not sure about adding "query" or filter functionality. This would require additional complexity from each backend (including said imagined ones) without adding apparent value. What would be a use case for having such filtering functionality in a `package_count` method?

File: pypiserver/_app.py
Comment (elfjes) [reply pypiserver/_app.py:333]: Yeah sure :)

Good catch on the `fname_and_hash`. Apparently no test caught that

File: pypiserver/backend.py
Comment (elfjes) [reply pypiserver/backend.py:71]: I appreciate the sentiment, but I'm not quite sure about restricting the interface to `Iterator`. `Backend.get_projects()` for example returns a `set`. It doesn't look right to me to have to return an `iter(set(...))`. Guiding people to have as much lazy iterables as possibl can also be achieved through other means, such as documentation.

File: pypiserver/backend.py
Comment (elfjes) [reply pypiserver/backend.py:108]: Yeah this extra check is quite redundant

File: pypiserver/backend.py
Comment (elfjes) [reply pypiserver/backend.py:67]: Yeah, good idea. Documentation has to be improved for the whole "implement your own backend"-flow in general :)

File: pypiserver/backend.py
Comment (elfjes) [reply pypiserver/backend.py:156]: Alas, no `Protocol`s in python<3.8. I see that I put the type_hint for cache_manager in the `__init__` assignment rather that the function signature, so I'm changing that. 

Did you mean to add a sort of CacheManager interface/ base implementation that different backends may consume and/or extend? This might be an idea, but I'm afraid that there is not going to be much common requirements between the different backends, since most of the logic will lie in cache invalidation, which is going to be very backend-specific

File: pypiserver/backend.py
Comment (elfjes) [reply pypiserver/backend.py:135]: I'm not too familiar with mypy and I'm not quite sure what you mean with this. A subclass is a type, right? Or do you mean something else?
I'll add the new file(s) to the `tox.ini` and see what needs be done :)

**Comment by elfjes** [comment]
Thanks for the review, much appreciated :)

> Got a chance to give this a quick overview this evening, and I really like the direction! I have an in-progress branch to replace all usage of the old config with the stuff from #339 with all tests passing. You can see that [here](https://github.com/pypiserver/pypiserver/compare/config-replace). It seems like that could mesh really well with the approach taken here. I think we would update the `RunConfig` to expose a `backend` attribute, rather than having it expose an `iter_packages` method.

Yeah I was also thinking of having a `--backend` attribute which picks a backend. See also `plugin.py` in which I philosophized a bit on how we can support selection and configuration of different backends

> > Simplify the PgkFile class. It currently consists of a number of attributes that don't necessarily belong with it, and not all attributes are aptly named (imho). I would like to minimalize the scope of PkgFile so that its only concern is being a data carrier between the app and the backends, and make its use more clear.
> 
> I agree with this goal! I don't feel strongly either way on whether it should be a part of this PR, but I definitely think it's worth exploring
Maybe it's best to give it it's own PR, so that we don't have to make this one larger than it already is :)

> > Add a PkgFile.metadata that backend implementations may use to store custom data for packages. For example the current PkgFile.root attribute is an implementation detail of the filestorage backends, and other Backend implementations should not be bothered by it.
> 
> Yeah, and having some access to metadata will also give us a route towards e.g. fully supporting PEP 503 and the `data_requires` attribute (a la #320).
Yeah I realized that there are two types of metadata that we must distinguish. One is the PEP503 metadata that we need for full support of this PEP. The other is geared towards extra data that a specific backend implementation may need to keep track of it's packages, such as a package `root` for a FileBackend, a bucket name for an imagined MultiBucketGCSBackend or a packages primary key in a database, and have a designated place to store such additional attributes. My thought initially was with the latter, but we need the former as well. 

> Regarding getting both of our streams of work going harmoniously, we can either get this merged, and then I can rebase and update on top of this, or vice-versa. I'll probably go ahead and open a PR with that branch tonight or tomorrow, just so that we can better look at how they might work together, but there's no rush to get that one in, and I'm happy to do the work of rebasing if we get this one in first.

Yeah sure, I'll finish up this PR and then create a new one for the additional work to ease the integration pain a little bit

File: pypiserver/_app.py
Comment (mplanchard) [reply pypiserver/_app.py]: Sorry, I just meant a count without any custom filtering, but I expressed that poorly. Where I was talking about a regex, I essentially just meant "count every file that looks like a package based on the filename"

File: pypiserver/backend.py
Comment (mplanchard) [reply pypiserver/backend.py:81]: Yeah, that seems reasonable. `Iterator` and `Iterable` can generally be consumed the same way. I just struggle to see any case where we'd want to be eager rather than lazy with getting packages. Even pulling from the local filesystem, it's probably better to be lazy to avoid being RAM hogs. I'm good with either approach though

File: pypiserver/backend.py
Comment (mplanchard) [reply pypiserver/backend.py:148]: Yep, I just meant that we would need to type these methods for mypy to be happy, and that mypy will enforce that they match the base class types to satisfy the liskov substitution principle

File: pypiserver/backend.py
Comment (mplanchard) [reply pypiserver/backend.py]: > Alas, no Protocols in python<3.8

This is true, but I think that if we want to use it, you can get it for earlier pythons by installing [typing-extensions](https://pypi.org/project/typing-extensions/).

> Did you mean to add a sort of CacheManager interface/ base implementation that different backends may consume and/or extend

Yep, that was the idea, so that you could potentially have like a `WatchdogCachingFileBackend` and a `WhateverOtherCachingFileBackend`, so long as the mangers share the same interface. It seems like all
we need for that at the moment is `listdir()` and `digest_file()`, so it might not be too bad to specify an abstract interface

**Comment by elfjes** [comment]
I've added a `--backend` option to the cli that by default automatically chooses between the caching and the simple file backend.

It took me a while to figure out how to exactly add the argument and make it work with the `Config.default_with_overrides` function that is used in the tests. Which got me thinking, if we want to add pluggable backends, these new backends may want to add new command line arguments. How would we deal with them in the currently design of the (Run)Config, which has fixed supported arguments? I guess we'd have to go back to some form of dynamic config objects, although that's probably a discussion for a different issue/PR

Speaking of pluggable backends, @mplanchard when you have time, could you give me your feedback on the proposal I made for plugins in `plugin.py`?

Also, if you're ok with the current status of this PR, we can merge it, so that I can start working on the other things in a separate PR.

**Comment by mplanchard** [comment]
> It took me a while to figure out how to exactly add the argument and make it work with the Config.default_with_overrides function that is used in the tests

There is definitely still room for improvement in making these obvious/easy to add to. 

> Which got me thinking, if we want to add pluggable backends, these new backends may want to add new command line arguments. How would we deal with them in the currently design of the (Run)Config, which has fixed supported arguments? I guess we'd have to go back to some form of dynamic config objects, although that's probably a discussion for a different issue/PR

I think the way we would approach this would be to have a dynamic portion of the config object, e.g. `backend_config`, which is a generic namespace. Plugins would need to provide some function to parse arguments and return a namespace, which we'd use to populate that property. Or something like that :shrug:

**Comment by elfjes** [comment]
yeah that could work. Or like a `**custom` argument to the config object that gets converted to a namespace. Or just kept as a dictionary

File: pypiserver/cache.py
Comment (mplanchard) [reply pypiserver/cache.py:20]: Ah I see you added this already!

File: pypiserver/config.py
Comment (mplanchard) [inline pypiserver/config.py:753]: Should we put this in the `_ConfigCommon` so that it's available for the update command too?

File: pypiserver/config.py
Comment (mplanchard) [inline pypiserver/config.py:527]: Should we move the `hash_algo` init argument to this common config?

File: pypiserver/core.py
Comment (mplanchard) [inline pypiserver/core.py:38]: thanks for this documentation!

File: pypiserver/core.py
Comment (mplanchard) [inline pypiserver/core.py:63]: Should this be `digester` to match the `__slots__`?

File: pypiserver/plugin.py
Comment (mplanchard) [inline pypiserver/plugin.py:24]: Can we pull this from `config.DEFAULTS`?

File: pypiserver/config.py
Comment (elfjes) [reply pypiserver/config.py:753]: Yes, that makes sense to me. I think the main reason I didn't do that yet, was because it currently requires `hash_algo` to be part of the same config and that one's on the RunConfig. But if you're fine with moving that to common, then I see no reason to not do that.

File: pypiserver/core.py
Comment (elfjes) [reply pypiserver/core.py:38]: Haha, I mostly put it there for my own sanity until I had the chance to refactor `PkgInfo`

File: pypiserver/core.py
Comment (elfjes) [reply pypiserver/core.py:63]: Both digest and digester are part of slots. I do see that I need to assign `self.digester=None` in `__init__` to be consistent

File: pypiserver/plugin.py
Comment (elfjes) [reply pypiserver/plugin.py:24]: So in my opinion, if we're doing this as plugins, then the package root is a config setting specific to the file backends. Which would also mean that the cli argument should not be part of the global config, but provided by the plugin. Which is why I added that definition there, and not take it from the global config

Comment (mplanchard) [reply pypiserver/plugin.py:24]: That does make sense. I just want to avoid needing to update this twice if we update it (although the likelihood that we'll ever change this is pretty low tbf)

File: pypiserver/config.py
Comment (mplanchard) [reply pypiserver/config.py:753]: Yep, now that we need it in both the run and update configs, we might as well move it

**Comment by elfjes** [comment]
Hi @mplanchard,

It's been a while, how are you? What do you think about the status of this PR? As far as I'm concerned it's ready to be merged. I've processed most, if not all of your feedback so far. Are there still any blocking things?

File: pypiserver/config.py
Comment (mplanchard) [inline pypiserver/config.py:610]: Dead code?

**Review by mplanchard — APPROVED** [review]
Sorry for being slow on this. Q4 and the beginning of Q1 have been crazy at my day job. Let's get it in!

**Comment by mplanchard** [comment]
Leaving it to you to merge. Let me know if you don't have permissions, and I'll hit the button instead

**Comment by dee-me-tree-or-love** [comment]
@mplanchard @elfjes if no rights to merge, I can do also! 
(sorry for my low participation lately as well... also quite overloaded with various tasks on the job. @mplanchard, what do you also think, actually is it okay for me to also be merging/approving when I get a moment for it, maybe I will be poking for your call upfront in the future still?)

**Comment by mplanchard** [comment]
> @mplanchard, what do you also think, actually is it okay for me to also be merging/approving when I get a moment for it, maybe I will be poking for your call upfront in the future still

Yeah I think so! For bigger stuff, we should probably both review, but for small things I'm okay with either of us approving and merging. And feel free to tag me for a review if there's anything at all you're uncertain about or would like me to weigh in on. If there's anything time sensitive, bug me in Zulip. It's harder for me to ignore real time chat than GH issues 🤪.

File: pypiserver/config.py
Comment (elfjes) [reply pypiserver/config.py]: Ah thanks, yeah I thought I might need to check that here, but it's actually picked up by argparse

**Comment by elfjes** [comment]
I removed that dead code you mentioned, so I guess it needs reapproval. I don't have a merge button so if any of you @mplanchard @dee-me-tree-or-love can merge it, I'd be much obliged :)

**Review by mplanchard — APPROVED** [review]

**Pull request merged**
